client/master_client.go

```diff
@@ -13,38 +16,64 @@ import (
 	"google.golang.org/grpc"
 )
 
-type MasterClient struct {
-	urls   []string
-	leader string
-	conn   closeable
+const dialTimeout = 5 * time.Second
+
+type clientHolder struct {
+	conn   closeableConnIface
 	client pb.MasterClient
 }
 
+type MasterClient struct {
+	urls    []string
+	leader  string
+	clients map[string]*clientHolder
+}
+
 func (c *MasterClient) init(ctx context.Context) error {
-	log.L().Logger.Info("dialing master", zap.String("leader", c.leader))
-	conn, err := grpc.DialContext(ctx, c.leader, grpc.WithInsecure(), grpc.WithBlock())
-	if err != nil {
-		return errors.ErrGrpcBuildConn.GenWithStackByArgs(c.leader)
+	log.L().Logger.Info("dialing master", zap.Strings("urls", c.urls))
+	for _, addr := range c.urls {
+		ctx, cancel := context.WithTimeout(ctx, dialTimeout)
+		conn, err := grpc.DialContext(ctx, addr, grpc.WithInsecure(), grpc.WithBlock())
+		if err != nil {
+			log.L().Warn("dial to one of server master failed", zap.String("addr", addr))
+			cancel()
+			continue
+		}
+		cancel()
```

Comment:
why invoke the cancel function here?

Reply:
This cancels the timeout context in L35, which is only used when dialing to server master.